### PR TITLE
improve the 'dimensions' tag rst rendering

### LINKED ISSUE #936: DOC BUG: NXsample field names have shape info in the manual
Some of the field names in the NXsample [documentation](https://manual.nexusformat.org/classes/base_classes/NXsample.html#nxsample) show shape information as part of the name.  The names are non-compliant with the naming rules.  This has led some people new to NeXus to define fields with the non-compliant names.  Needs reformatting.

2121-07-16 NXsample field names

### PR DESCRIPTION
**Pull request by woutdenolf**
Closes #936 This does not change the meaning of the `dimensions` xml tag but rather the rst rendering. Problems with the original rendering:
* some users were using for example `"field[n]"` as the HDF5 dataset name instead of `"field"`
* rendering sometimes gave the wrong idea about rank symbol and/or dimensions symbols (e.g. the DATA fiield of NXdata, see bellow)

These are the different dimensions that can occur:

## Fixed rank

```xml
<field name="data" type="NX_INT" signal="1">
  <dimensions rank="3">
    <dim index="1" value="nP" />
    <dim index="2" value="xDim" />
    <dim index="3" value="yDim" />
  </dimensions>
</field>
```

[NXscan example](https://manual.nexusformat.org/classes/applications/NXscan.html#nxscan-entry-instrument-detector-data-field)

Old rendering

image

New rendering

image

## Variable rank because of optional dimensions

```xml
<field name="data" type="NX_NUMBER" recommended="true">
    <doc>
        For a dimension-2 detector, the rank of the data array will be 3.
        For a dimension-3 detector, the rank of the data array will be 4.
        This allows for the introduction of the frame number as the
        first index.
    </doc>
    <dimensions rank="dataRank">
        <dim index="1" value="nP" />
        <dim index="2" value="i" />
        <dim index="3" value="j" />
        <dim index="4" value="k" required="false"/>
    </dimensions>
</field>
```

[NXmx example](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-data-field)

Old rendering

image

New rendering

image

## Variable rank because no dimensions specified

```xml
<field name="DATA" type="NX_NUMBER" nameType="any">
   <dimensions rank="dataRank">
	<doc>
		The rank (``dataRank``) of the ``data`` must satisfy
		``1 &lt;= dataRank &lt;= NX_MAXRANK=32``.  
		At least one ``dim`` must have length ``n``.
	</doc>
	<dim index="0" value="n"><!-- index="0": cannot know to which dimension this applies a priori --></dim>
   </dimensions>
</field>
```

The legacy way above is still supported but you could also do this

```xml
  <dimensions rank="dataRank">
  </dimensions>
```

[NXdata example](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-data-field)

image

Old rendering

image

New rendering

image

## Same rank and dimensions as another field

```xml
<field name="group_index" type="NX_INT">
	<doc>
	An array of unique identifiers for detectors or groupings 
	of detectors.
	
	Each ID is a unique ID for the corresponding detector or group
	named in the field group_names.  The IDs are positive integers
	starting with 1. 
	</doc>
       <dimensions><dim index="1" value="i"/></dimensions>
</field>
<field name="group_parent" type="NX_INT">
	<doc>
	An array of the hierarchical levels of the parents of detectors
	or groupings of detectors.
	
	A top-level grouping has parent level -1.</doc>
	<dimensions><dim index="1" ref="group_index"/></dimensions>
</field>
```

[Nxdetector_group example](https://manual.nexusformat.org/classes/base_classes/NXdetector_group.html#nxdetector-group-group-parent-field)

Old rendering

image

New rendering

image

### CODE REVIEW

File: utils/nxdl2rst.py [hunk 1]
```diff
@@ -201,20 +201,106 @@ def get_required_or_optional_text(node, use_application_defaults):
...
+            # the rank is variable because of one or more
+            # dimensions are optional
+            rank = len(dims)
+        else:
+            # the rank is variable because no dimensions
+            # were specified
+            rank = "."
+
+    if dims_from_field:
+        return ' (Rank: %s, Dimensions:%s)' % (rank, dims)
```
Comment (benajamin) [inline utils/nxdl2rst.py:298]: please add a space between "Dimensions:" and "%s" both here and on line 301.

**Review by benajamin — APPROVED** [review]
Looks like progress to me!

**Comment by woutdenolf** [comment]
This change is probably going to expose improper usage of the `dimensions` tag. For example in `NXsensor` to `ref` is used in the wrong way:

image

**Comment by woutdenolf** [comment]
Ok it seems I misinterpreted the `ref=`. I'll fix it ...

**Comment by prjemian** [comment]
I like these changes.  The format can be more clear, as the metadata about each field is represented in different ways (it is added by different parts of the nxdl2rst converter).  As shown:

**value (Rank:1, Dimensions: [n])**: (optional) [NX_FLOAT](https://example.link) (units=[NX_ANY](https://example.link))

It is important to see the field name clearly and first know if it is optional.  Let's move the metata a bit, putting the name of the field first and clearly separated from the metadata.  Can the new metadata be inserted after the data type?

**value**: (optional) [NX_FLOAT](https://example.link) (Rank:1, Dimensions: [n]) (units=[NX_ANY](https://example.link))

**Comment by woutdenolf** [comment]
I fixed the mistake with `<dim ... ref="">`: it means the rank and dimensions are the same a the field that is referenced.

**Comment by woutdenolf** [comment]
@prjemian I modified to match your suggestion

```
value: (optional) [NX_FLOAT](https://example.link/) (Rank:1, Dimensions: [n]) (units=[NX_ANY](https://example.link/))
```

**Comment by prjemian** [comment]
The new screen shots use different variables than shown in the example NXDL `<dimensions />` block.

**Review by prjemian — APPROVED** [review]
LGTM!

File: utils/nxdl2rst.py [hunk 2]
```diff
@@ -326,8 +405,8 @@ def printFullTree(ns, parent, name, indent, parent_path):
         print( '%s.. index:: %s (field)\n' %
                ( indent, index_name ) )
         print(
-            '%s**%s%s**: %s%s%s\n' % (
-                indent, name, dims, optional_text, fmtTyp(node), fmtUnits(node)
+            '%s**%s**: %s%s%s%s\n' % (
```
[the remark this answers is not in the document]
Comment (woutdenolf) [reply utils/nxdl2rst.py:408]: Yeah sure

**Comment by woutdenolf** [comment]
> The new screen shots use different variables than shown in the example NXDL `<dimensions />` block.

Yeah sorry, I was a bit lazy. I fixed that. The xml and the images should match.

**Comment by woutdenolf** [comment]
@prjemian I'm off for today. You can merge it or wait the other reviewers.

**Comment by prjemian** [comment]
Thanks! I can wait to merge for a day, to allow others a chance to review.

**Pull request merged**